File: leetcode/top150/138.cpp
/*
// Definition for a Node.
class Node {
public:
    int val;
    Node* next;
    Node* random;

    Node(int _val) {
        val = _val;
        next = NULL;
        random = NULL;
    }
};
*/

class Node {
  public:
    int val;
    Node* next;
    Node* random;

    Node(int _val) : val(_val), next(nullptr), random(nullptr) {}
    Node(int _val, Node* _next, Node* _random)
        : val(_val), next(_next), random(_random) {}
};

class Solution {
  public:
    Node* copyRandomList(Node* head) {
        Node* dummy = new Node(0, head, nullptr);
        while (head) {
            head->next = new Node(head->val, head->next, nullptr);
            head = head->next->next;
        }
        head = dummy->next;
        while (head) {
            if (head->random) {
                head->next->random = head->random->next;
            }
            head = head->next->next;
        }
        head = dummy;
        while (head->next) {
            Node* nxt = head->next;
            head->next = nxt->next;
            head = nxt;
        }
        return dummy->next;
    }
};

File: leetcode/hot100_python/138.py
from typing import Optional


class Node:
    def __init__(self, val: int, next: 'Node' = None, random: 'Node' = None):
        self.val = val
        self.next = next
        self.random = random

class Solution:
    def copyRandomList(self, head: 'Optional[Node]') -> 'Optional[Node]':
        if head is None:
            return None
        cur = head
        while cur:
            cur.next = Node(cur.val, cur.next)
            cur = cur.next.next
        cur = head
        while cur:
            if cur.random:
                cur.next.random = cur.random.next
            cur = cur.next.next
        new_head = head.next
        cur = head
        while cur.next:
            tmp = cur.next
            cur.next = cur.next.next
            cur = tmp
        return new_head

